go/clustermetadata/topopublish/topopublish.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2025 Supabase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package topopublish manages the publishing of components to the topo.
+package topopublish
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"sync"
+	"time"
+
+	"github.com/multigres/multigres/go/servenv"
+	"github.com/multigres/multigres/go/tools/timertools"
+)
+
+// TopoPublisher contains the metadata of the component being published.
+type TopoPublisher struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+	logger *slog.Logger
+
+	unpublish func(ctx context.Context) error
+}
+
+// Publish publishes the component using the publish function. If the publish function
+// returns an error, it will be retried with exponential backoff until successful.
+// The alarm will be invoked with the latest error message during retries. If the
+// publishing succeeds, the alarm will be invoked with an empty string.
+func Publish(publish func(ctx context.Context) error, unpublish func(ctx context.Context) error, alarm func(string)) *TopoPublisher {
```

Comment:
Idea on naming: Something like Register seems clearer in this context than publish.